src/rpc/doc.cpp

```diff
@@ -0,0 +1,284 @@
+    result += "\n";
+    return result;
+}
+
+RPCDocTableRow::RPCDocTableRow(const std::string& code)
+    : m_code(code)
+{
+}
+
+RPCDocTableRow::RPCDocTableRow(const std::string& code, const std::string& description)
+    : m_code(code), m_description(description)
+{
+}
+
+RPCDocTableRow::RPCDocTableRow(const std::string& code, const std::initializer_list<std::string>& types, const std::string& description)
+    : m_code(code), m_types(types), m_description(description)
+{
+}
+
+std::string const& RPCDocTableRow::Code() const
+{
+    return m_code;
+}
+
+std::vector<std::string> const& RPCDocTableRow::Types() const
+{
+    return m_types;
+}
+
+std::vector<std::string> RPCDocTableRow::DescriptionLines() const
+{
+    std::vector<std::string> res;
+    boost::split(res, m_description, boost::is_any_of("\n"));
+    return res;
+}
+
+size_t RPCDocTable::PrefixLength() const
+{
+    size_t max = 0;
+    for (auto const& row : m_rows) {
+        size_t prefix = row.Code().length() + 2;
+        if (prefix > max) {
+            max = prefix;
+        }
+    }
+    return max;
+}
+
+
+RPCDocTable::RPCDocTable(const std::string& name)
+    : m_name(name)
+{
+}
+
+void RPCDocTable::AddRow(const RPCDocTableRow& row)
+{
+    m_rows.push_back(row);
+}
+
+std::string RPCDocTable::AsText() const
+{
+    std::string res = "";
+    res += m_name;
+    res += ":\n";
+
+    size_t prefixLen = PrefixLength();
+    for (auto const& row : m_rows) {
+        std::string code = row.Code();
+        res += row.Code();
+        auto lines = row.DescriptionLines();
+        bool firstLine = true;
+        for (auto const& line : lines) {
+            size_t spaces;
+            if (firstLine) {
+                spaces = prefixLen - code.length();
+            } else {
+                spaces = prefixLen;
+            }
+            res += std::string(spaces, ' ');
+
+            if (firstLine) {
+                auto types = row.Types();
+                if (!types.empty()) {
+                    bool firstType = true;
+                    res += "(";
+                    for (auto const& type : types) {
+                        if (!firstType) {
+                            res += ", ";
+                        }
+                        res += type;
+                        firstType = false;
+                    }
+                    res += ") ";
+                }
+            }
+            res += line;
+            res += "\n";
+            firstLine = false;
+        }
+    }
+    return res;
+}
+
+RPCDoc::RPCDoc(std::string methodName, std::string firstArguments)
+    : m_methodName(methodName), m_firstArguments(firstArguments)
+{
+}
+
+RPCDoc::RPCDoc(std::string methodName)
+    : m_methodName(methodName), m_firstArguments("")
+{
+}
+
+RPCDoc& RPCDoc::Desc(const std::string& description)
+{
+    m_description = description;
+    return *this;
+}
+
+RPCDoc& RPCDoc::Table(const std::string& name)
+{
+    m_tables.push_back(RPCDocTable(name));
```

Comment:
`emplace_back` instead?

Reply:
I have to admit, I don't really understand the difference between emplace_back and push_back

What good does it bring here?

Reply:
...but online wisdom seems to be that emplace_back is faster, because it doesn't create temporary objects, so... ok :)